File: backend/models/trades.py
from __future__ import annotations

from pydantic import BaseModel, Field


class TradeAttemptResult(BaseModel):
    rank: int
    underlying_symbol: str
    success: bool
    error: str | None = None
    trade_id: str | None = None
    order_status: str | None = None


class AutonomousExecutionResult(BaseModel):
    executed: bool
    selected_rank: int | None = None
    trade_id: str | None = None
    underlying_symbol: str | None = None
    attempts: list[TradeAttemptResult] = Field(default_factory=list)
    message: str

File: frontend/src/types/trades.ts
export interface TradeAttemptResult {
  rank: number;
  underlying_symbol: string;
  success: boolean;
  error?: string;
  trade_id?: string;
  order_status?: string;
}

export interface AutonomousExecutionResult {
  executed: boolean;
  selected_rank?: number;
  trade_id?: string;
  underlying_symbol?: string;
  attempts: TradeAttemptResult[];
  message: string;
}

export type RankExecutionStatus = "pending" | "attempting" | "succeeded" | "failed" | "skipped";

export function rankStatusFromResult(
  rank: number,
  result: AutonomousExecutionResult | null,
  isExecuting: boolean,
): RankExecutionStatus {
  if (!result) {
    return isExecuting && rank === 1 ? "attempting" : "pending";
  }

  const attempt = result.attempts.find((a) => a.rank === rank);
  if (!attempt) {
    if (result.executed && result.selected_rank != null && rank > result.selected_rank) {
      return "skipped";
    }
    return "pending";
  }
  if (attempt.success) return "succeeded";
  return "failed";
}
